### ISSUE TITLE
Shelly Dimmer support possible ?

### ISSUE THREAD

**Issue by andyblac** [mongoose-os-apps/shelly-homekit#14]
as title do you think Shelly Dimmer would possible ?

**Comment by rojer** [comment]
in principle, yes. but it has secondary microcontroller, which is an additional complication.

**Comment by andyblac** [comment]
I could do with it, would you be willing to take a look if I send you one ?

**Comment by andyblac** [comment]
could this help ? https://github.com/alexryd/homebridge-shelly

**Comment by rojer** [comment]
no, that's not relevant.

i would not recvommend starting with dimmer. implementing input as push button would be easier and get you acquainted with what implementing a HAP service entails.

**Comment by andyblac** [comment]
an a dimmer act as on/off too ?

**Comment by rojer** [comment]
what's the use for the dimmer without, you know, dimming functionality?
i think we should hold off declaring support for shelly dimmer until dimming is implemented.
in terms of HAP, that would mean implementing a lightbulb service, which supports variable brightness.

**Comment by andyblac** [comment]
so if I purchased a shelly dimmer I could try adding as a switch to start ?

**Comment by kzyapkov** [comment]
The dimmer is indeed not the device to start with. If lighting is your thing, try the LED drivers, they should be much simpler, using PWM for output control.

As for the dimmer -- we plan on releasing info about the UART protocol used to control the small MCU which does the actual dimming, but no ETA.

**Comment by andyblac** [comment]
ahhhh. atm I'm using Philips Hue, but tbh it is unreliable, found your shelly code match more so and it is instant, I was hoping to replace my Hue with Shelly's.

**Comment by kzyapkov** [comment]
please drop a note to developers@shelly.cloud, we may be able to help

**Comment by andyblac** [comment]
atm I have zero knowledge of HAP or MOS, I am a python coder. mOS looks vastly different, I am trying to read through this git code to try to get an understanding of it.

**Comment by rojer** [comment]
right, so dimmer is not a good start. better pick a simpler issue to tackle.
for HAP, you bible is `HAP-Specification-Non-Commercial-Version.pdf`. you can download it from [here](https://developer.apple.com/homekit/). you'll need to login with apple id but it's free of charge.
get familiar with the concepts - controller, accessory, service, characteristic, event.
then read existing code to understand how services are defined and characteristics and events managed. fortunately, it's not that much - all of the "switch" service implemented by shellies is contained [here](https://github.com/mongoose-os-apps/shelly-homekit/blob/master/src/shelly_sw_service.c).

in terms of coding - yeah, it's the good old C, not Python. not much do be done about it, i'm afraid. either you pick it up, or you won't be able to do much.

**Comment by andyblac** [comment]
thanks, I have some homework to do 😄

**Comment by sebastianheise** [comment]
Hi guys! How are you doing? I've caught up on this thread and had a look at the HAP. 
From my understanding a _dimmer_ would be implemented using the _brightness_ characteristic. Correct? So instead of a switch, the device would need to authenticate as a _light bulb_ service. Is that assumption correct? Unfortunately I also don't have any experience in HAP and MOS, but it sounds like a fun project. If I can help with anything I'm all keen. Just give me a yell. Would be awesome to have support for the Shelly dimmer!

**Comment by rojer** [comment]
yes, that is correct.

**Comment by Olly68** [comment]
@rojer 
Do you have plans for Shelly Dimmer 2 at the moment?
Or is time still your problem?

It would be great if you can find a solution👌🏻😀

**Comment by ree109** [comment]
Would this help with the dimmer - https://github.com/curla92/LED-Dimmer-Homekit

**Comment by vanrri** [comment]
Anyone tried it yet?

**Comment by andyblac** [comment]
Just so you don't get Bored!!! HAHA, Can This Be Next Please!!!

**Comment by rojer** [comment]
yes, i think i'll tackle either dimmer or rgbw2 next. we'll see.

**Comment by andyblac** [comment]
cool cheers!!!

**Comment by andyblac** [comment]
> yes, i think i'll tackle either dimmer or rgbw2 next. we'll see.

do you need any devices ?, I don't mind donating one of each for you to work with.

**Comment by rojer** [comment]
no, i'm all set. i've had a couple for a while now.

**Comment by andyblac** [comment]
ok np.

**Comment by freddair** [comment]
+1 for Dimmer next :)

**Comment by rojer** [comment]
yes, dimmer is next (after garage door is out), in particular dimmer2, which is what i have for testing.

**Comment by andyblac** [comment]
> yes, dimmer is next (after garage door is out), in particular dimmer2, which is what i have for testing.

if I can help with debug info for Dimmer1 let me know.

**Comment by freddair** [comment]
+1 for Dimmer1

**Comment by Leo-W536** [comment]
+1 for Dimmer 1 👍🏻

**Comment by andyblac** [comment]
i now have dimmer2, so can help testing that too 😄

**Comment by ToblerCGN** [comment]
I got two dimmers 2. I'd like to help, too!

**Comment by Przemyslaw-Wosko** [comment]
@rojer can we implemnet shelly dimmer as light bulb instead? it has characteristics to on / off and also dimm 
i will try to update some code, and rise PR. 

What do you think?

**Comment by rojer** [comment]
that's exactly the plan, dimmer will be represented as a lightbulb HAP service with "brightness" characteristic.

**Comment by Przemyslaw-Wosko** [comment]
Have you started work somewhere? i've noticed it will be lot of work, since everything relays on Switch 
I'd love to collaborate instead of writing whole stuff from scratch

**Comment by rojer** [comment]
the biggest chunk of work will be implementing the UART protocol to communicate with the MCU that does the dimming.
adding a HAP service, will be the easy part.
no, i haven't written any code yet, but i have test setup and have looked into the protocol details, i have a reasonable grasp of it now.
will probably get something written this weekend.

**Comment by rojer** [comment]
well, this did not go to plan at all.
i was having issues getting ,y firmware to run after OTA, and while messing with serial connections, briefly had both non-isolated serial connected while i turned on the mains power... BOOM! sparks, smoke, lights go out.
when the smoke cleared, the casualties are: 2 of 2 dimmers, usb hub and, crucially, my main laptop.
nothing much to show for it. not a good day...
but i'll be back :)

**Comment by pivale** [comment]
> well, this did not go to plan at all.
> i was having issues getting ,y firmware to run after OTA, and while messing with serial connections, briefly had both non-isolated serial connected while i turned on the mains power... BOOM! sparks, smoke, lights go out.
> when the smoke cleared, the casualties are: 2 of 2 dimmers, usb hub and, crucially, my main laptop.
> nothing much to show for it. not a good day...
> but i'll be back :)

damn :( sorry to hear that!

**Comment by freddair** [comment]
@rojer glad to understand that yourself have not been harmed though! I have a few stock Dimmer1 laying around, so let me know if and how I can provide you with one!

**Comment by andyblac** [comment]
@rojer Dam, thats not good, the pesky none isolated serial is dangerous. Just glad to here your OK though, kit can be replaced. I think we need to get you some donations in to help you replace that laptop. One from me on the way!!

**Comment by Przemyslaw-Wosko** [comment]
@rojer thanks for effort, i didn't knew this could go soo bad and i feel sorry for your laptop, i hope it wasn't something extremally expensive

**Comment by freddair** [comment]
@rojer do you already have the needed Dimmer UART specification? Otherwise maybe the sample config of HAA firmware could help:

`{"t":30,"d":0.1,"b":[{"g":12,"i":1,"p":0},{"g":12,"t":0}],
"0":{"u":[{"v":"010202060000f401000000ff04"}]},
"y0":[{"v":1,"0":{"u":[{"v":"010202060a00f4010000010904"}]}},
{"v":5,"0":{"u":[{"v":"010202063200f4010000013104"}]}},
{"v":10,"0":{"u":[{"v":"010202066400f4010000016304"}]}},
{"v":15,"0":{"u":[{"v":"010202069600f4010000019504"}]}},
{"v":20,"0":{"u":[{"v":"01020206c800f401000001c704"}]}},
{"v":25,"0":{"u":[{"v":"01020206fa00f401000001f904"}]}},
{"v":30,"0":{"u":[{"v":"010202062c01f4010000012c04"}]}},
{"v":35,"0":{"u":[{"v":"010202065e01f4010000015e04"}]}},
{"v":40,"0":{"u":[{"v":"010202069001f4010000019004"}]}},
{"v":45,"0":{"u":[{"v":"01020206c201f401000001c204"}]}},
{"v":50,"0":{"u":[{"v":"01020206f401f401000001f404"}]}},
{"v":55,"0":{"u":[{"v":"010202062602f4010000012704"}]}},
{"v":60,"0":{"u":[{"v":"010202065802f4010000018b04"}]}},
{"v":65,"0":{"u":[{"v":"010202068a02f4010000010904"}]}},
{"v":70,"0":{"u":[{"v":"01020206bc02f401000001bd04"}]}},
{"v":75,"0":{"u":[{"v":"01020206ee02f401000001ef04"}]}},
{"v":80,"0":{"u":[{"v":"010202060032f4010000013104"}]}},
{"v":85,"0":{"u":[{"v":"010202065203f4010000015404"}]}},
{"v":90,"0":{"u":[{"v":"010202068403f4010000018604"}]}},
{"v":95,"0":{"u":[{"v":"01020206b603f401000001b804"}]}},
{"v":100,"0":{"u":[{"v":"01020206e803f401000001ea04"}]}}]}`

**Comment by rojer** [comment]
@freddair that's a neat hack. i'd like to have something more sophisticated than that, though.

**Comment by freddair** [comment]
of course! 👍

**Comment by Przemyslaw-Wosko** [comment]
@rojer done is better than perfect :D wouldn't that be sufficient as proof of concept for dimmers? it could be upgraged in future

**Comment by andyblac** [comment]
> @rojer done is better than perfect :D wouldn't that be sufficient as proof of concept for dimmers? it could be upgraged in future

imo, HAA is not the way to go, it far to hacky for my liking, i would rather it done correctly in the place.

**Comment by freddair** [comment]
> > @rojer done is better than perfect :D wouldn't that be sufficient as proof of concept for dimmers? it could be upgraged in future
> 
> imo, HAA is not the way to go, it far to hacky for my liking, i would rather it done correctly in the place.

Yeah I tried HAA on my Shelly Dimmer 1 before and it was less than perfect - from time to time I had issues with on/off blinking lights and input confusions / crashes... but I thought that UART commands to the dimmer chip would likely not have been the cause.

**Comment by andyblac** [comment]
i did find a github thread that gave @rojer the info he needed, we just need to be patient.

https://github.com/arendst/Tasmota/issues/6914#issuecomment-562109267

**Comment by rojer** [comment]
i don't intend to polish every last bit for the first iteration, but having just a couple fixed commands to bang out is a bit too low a bar :)
since no actual progress is expected until i get my laptop back and a couple test devices,  let me dump my thoughts so far instead :)

so, the way dimmer works is quite complicated. it has a slave mcu that does the PWM and communication is over UART. the same UART that is normally used for flashing and debug, so that's complication number 1.
second, the protocol is not documented. there are multiple commands.
third, the slave MCU formware is tightly coupled to the stock ESP firmware. STM firmware is shipped with the ESP firmware and is stored on the filesystem. ESP will flash it first thing after boot if it detects version mismatch.
finally, there is the calibration procedure - calibration data is stored on the filesystem, and is sent to the STM on boot.

so, there's quite a lot of ground to cover here to get to feature parity with the stock firmware, but we can do it in stages.

stage 0, early testing:
 * implement uart protocol
 * only support setting brightness, nothing else
 * ignore calibration data
 * ignore STM firmware

this is essentially what HAA bunch of strings does, but a little more sophisticated - we'll actually use proper sequence numbers and check responses.

stage 1:
 * support uploading existing calibration data, when calibration was preformed on stock firmware.
i think this could be first release. we'll ask user to calibrate usinng stock fw, then flash. if they need to recalibrate, they'll have to revert to stock, calibrate, then flash homekit fw again.

stage 2:
 * add calibration procedure to homekit fw - perform calibration sequence, record data

stage 3:
 * add STM firmware update, so we always know what version STM is running

that's the path i see. of course, i stumbled straight out of the gate - i didn't even get as far as running my code on the ESP before everything went up in smoke :)

**Comment by freddair** [comment]
@rojer great path! I would love to help in beta testing stage 0.

And I would also love to start with a simple switch implementation (light on/100% and off/0%), as the dimmer is already installed behind the wall switch and I fiddle with it since ages (HAA, home bridge-shelly didn't work either)... 

I am currently on stock fw again and really can't wait to have a working basic HomeKit implementation.

Let me know how I can send you one of my spare Dimmer1 :)

**Comment by rojer** [comment]
@freddair i'm focusing on dimmer2 at first, as that is the current product for sale. that said, extending support to dimmer1 is also planned, so if you can send me a device to test with, it'd be great. sned me an email to rojer@rojer.me and i'll give you the address.

**Comment by freddair** [comment]
@rojer done. I hoped that Dimmer 1 and 2 would be the same fw wise but yeah... i can('t) wait a few more weeks ;-)

**Comment by andyblac** [comment]
> @rojer done. I hoped that Dimmer 1 and 2 would be the same fw wise but yeah... i can('t) wait a few more weeks ;-)

could always setup homebridge, that what i doing atm, Dimmer 1/2 work ok, just dont even bother trying RGBW2 it just crashes the device.

**Comment by freddair** [comment]
> could always setup homebridge, that what i doing atm, Dimmer 1/2 work ok, just dont even bother trying RGBW2 it just crashes the device.

unfortunately home bridge-shelly did not work for me as for some reason it did not discover the dimmer1 (it's the only stock shelly I use)... https://github.com/alexryd/homebridge-shelly/issues/214

**Comment by bdshet** [comment]
> could always setup homebridge, that what i doing atm, Dimmer 1/2 work ok
@andyblac read from top to bottom — love to see the work you all are doing for native Shelly Dimmer HK support. In the meantime, what plugin are you using on HomeBridge?

**Comment by andyblac** [comment]
> > could always setup homebridge, that what i doing atm, Dimmer 1/2 work ok
> 
> @andyblac read from top to bottom — love to see the work you all are doing for native Shelly Dimmer HK support. In the meantime, what plugin are you using on HomeBridge?

https://github.com/alexryd/homebridge-shelly#readme

**Comment by andyblac** [comment]
> > could always setup homebridge, that what i doing atm, Dimmer 1/2 work ok, just dont even bother trying RGBW2 it just crashes the device.
> 
> @andyblac unfortunately home bridge-shelly did not work for me as for some reason it did not discover the dimmer1 (it's the only stock shelly I use)... [alexryd/homebridge-shelly#214](https://github.com/alexryd/homebridge-shelly/issues/214)

the new firmware 1.8.x is very picky on networks due CoAP, you could try reverting to version 1.7.0.

**Comment by bdshet** [comment]
@andyblac that's great advice. Thank you!

**Comment by Przemyslaw-Wosko** [comment]
@andyblac @rojer i'm waiting for this feature, since i have wall to finish on my flat :D 

Guys, what are changes, that this firmware will not require physical access to dimmer after this feature will be finished?

**Comment by andyblac** [comment]
> @andyblac @rojer i'm waiting for this feature, since i have wall to finish on my flat :D
> 
> Guys, what are changes, that this firmware will not require physical access to dimmer after this feature will be finished?

it shouldn't, but i would not try any beta versions, if any are released, wait till it's complete.

**Comment by rojer** [comment]
right, flashing will be done OTA, as with other devices, and won't require physical access. but as Andy said, staying away from early versions would be prudent.

**Comment by Przemyslaw-Wosko** [comment]
@rojer @andyblac thank you for advices!

**Comment by RedAnon** [comment]
So how far/close are we to at least some beta quality FW?

**Comment by rojer** [comment]
not very close, i'm afraid. i am going to release 2.6 with 1L support and a bunch of small features, then get back to dimmer when my main development laptop is back from repairs (should be soon).

**Comment by pivale** [comment]
Will there be an option to use rotary switches with dimmer?

**Comment by NotClear** [comment]
Any news about the Shelly dimmers?

**Comment by rojer** [comment]
sorry, no progress yet. been focusing on stability lately, plus busy with other things mongoose-os

**Comment by ToblerCGN** [comment]
Hi guys! Any news about dimmer support? Would love to see that.

**Comment by mrmaximas** [comment]
for me too. thanks/

**Comment by dwh2427** [comment]
And me - can we put some $ support behind this and whoever makes it happen in 30 days gets it?

Sent via Superhuman iOS ( https://sprh.mn/?vip=dave@loved.com )

**Comment by andyblac** [comment]
> And me - can we put some $ support behind this and whoever makes it happen in 30 days gets it? Sent via Superhuman iOS ( https://sprh.mn/?vip=dave@loved.com )
> […](#)

none us do this for money, it not about money, we do it in our free time, unfortunately it will happen when it does. i can't say when.

**Comment by dwh2427** [comment]
I appreciate that, it’s amazing when everything just works and really appreciate the passion behind it :)

Sent via Superhuman iOS ( https://sprh.mn/?vip=dave@loved.com )

**Comment by dwh2427** [comment]
But having said that, maybe people would be incentivized to jump on board with the work if there were some incentives so that’s why I thought I’d throw it out there

Sent via Superhuman iOS ( https://sprh.mn/?vip=dave@loved.com )

**Comment by freddair** [comment]
I'm also desperately waiting for my Dimmers to finally get reliable HomeKit support (tried HAA and homebridge-shelly but no...)

**Comment by andyblac** [comment]
> I'm also desperately waiting for my Dimmers to finally get reliable HomeKit support (tried HAA and homebridge-shelly but no...)

homebridge running ok here. try sticking to v1.1.6

**Comment by rojer** [comment]
i've been procrastinating on this for too long. unfortunately, i don't have an ETA still, this needs a significant chunk of work and i just couldn't find time and/or energy to put it in.

**Comment by freddair** [comment]
@rojer that's totally understandable. You want to do it right in the first place.

> homebridge running ok here. try sticking to v1.1.6

@andyblac you mean homebridge-Shelly v1.1.6 or stock fw?

**Comment by andyblac** [comment]
> @rojer that's totally understandable. You want to do it right in the first place.
> 
> > homebridge running ok here. try sticking to v1.1.6
> 
> @andyblac you mean homebridge-Shelly v1.1.6 or stock fw?

sorry hb v 1.1.6

**Comment by jotzet79** [comment]
> I'm also desperately waiting for my Dimmers to finally get reliable HomeKit support (tried HAA and homebridge-shelly but no...)

Be sure to disable „IGMP Snooping“ on your router and update to the latest Shelly FW for the Dimmer. This setup is working (since yesterday :) )

**Comment by andyblac** [comment]
> > I'm also desperately waiting for my Dimmers to finally get reliable HomeKit support (tried HAA and homebridge-shelly but no...)
> 
> Be sure to disable „IGMP Snooping“ on your router and update to the latest Shelly FW for the Dimmer. Together with homebridge my setup is working (well at least since yesterday :) )

disabling IGMP snooping is bad idea imo. it will cause all mDNS traffic to go everwhere, not just the ports that ask for it. @jotzet79 do you use a managed switch ?

**Comment by freddair** [comment]
> sorry hb v 1.1.6

I'm using HOOBS 3 Version | 3.3.2...

**Comment by jotzet79** [comment]
> disabling IGMP snooping is bad idea imo. it will cause all mDNS traffic to go everwhere, not just the ports that ask for it. @jotzet79 do you use a managed switch ?

@andyblac by "managed" you mean a Cable / DSL router, which is managed by the Internet provider? Or do you mean sth different?

**Comment by andyblac** [comment]
> > disabling IGMP snooping is bad idea imo. it will cause all mDNS traffic to go everwhere, not just the ports that ask for it. @jotzet79 do you use a managed switch ?
> 
> @andyblac by "managed" you mean a Cable / DSL router, which is managed by the Internet provider? Or do you mean sth different?

no, i mean following
> Managed switches have one or more methods to modify the operation of the switch. Common management methods include: a command-line interface (CLI) accessed via serial console, telnet or Secure Shell, an embedded Simple Network Management Protocol (SNMP) agent allowing management from a remote console or management station, or a web interface for management from a web browser. Examples of configuration changes that one can do from a managed switch include: enabling features such as Spanning Tree Protocol or port mirroring, setting port bandwidth, creating or modifying virtual LANs (VLANs), etc. Two sub-classes of managed switches are smart and enterprise managed switches.

**Comment by freddair** [comment]
Any news on this?

**Comment by rojer** [comment]
not yet, sorry

**Comment by andyblac** [comment]
dimmer is one of the hardest to support, due the hardware that if not done correctly it can be dangerous (it can overheat and melt) if the dimmer is used in wrong mode with light type (leading edge vs trailing edge).

IMO this is one device that should not be rushed.

**Comment by zechn** [comment]
Support for the shelly dimmer 2 would really be great, just bought some an then realized they are not supportes yet...

**Comment by Partymitch** [comment]
Hi. Just to inform you. I bought a Shelly Dimmer 2 with Homekit fimrware from this site: https://www.voltio.cz/en/dimmable/548-shelly-dimmer.html  .
It works fine, but it's not possible to calibrate. The lights are flackiring between 30 and 50 %. So the have realized a firmware. Perhaps you can get one to read out it ;-)

**Comment by BlancoDeMallloca** [comment]
> Hi. Just to inform you. I bought a Shelly Dimmer 2 with Homekit fimrware from this site: https://www.voltio.cz/en/dimmable/548-shelly-dimmer.html .

I also bought the Shelly RGBW from voltio.cz
Works pretty good. There is no web interface once it is set up. So you need to go back to set-up process (access point mode) every time you want to change settings.

**Comment by andyblac** [comment]
> Hi. Just to inform you. I bought a Shelly Dimmer 2 with Homekit fimrware from this site: https://www.voltio.cz/en/dimmable/548-shelly-dimmer.html .
> It works fine, but it's not possible to calibrate. The lights are flackiring between 30 and 50 %. So the have realized a firmware. Perhaps you can get one to read out it ;-)

be very careful if it's not correctly the dimmer can overheat and burn / melt, and without calibrations seem very basicly done.

**Comment by andyblac** [comment]
> > Hi. Just to inform you. I bought a Shelly Dimmer 2 with Homekit fimrware from this site: https://www.voltio.cz/en/dimmable/548-shelly-dimmer.html .
> 
> I also bought the Shelly RGBW from voltio.cz
> Works pretty good. There is no web interface once it is set up. So you need to go back to set-up process (access point mode) every time you want to change settings.

try 2.9.0-alpha2 it now has basic RGBW2 support.

**Comment by Partymitch** [comment]
> > Hi. Just to inform you. I bought a Shelly Dimmer 2 with Homekit fimrware from this site: https://www.voltio.cz/en/dimmable/548-shelly-dimmer.html .
> > It works fine, but it's not possible to calibrate. The lights are flackiring between 30 and 50 %. So the have realized a firmware. Perhaps you can get one to read out it ;-)
> 
> be very careful if it's not done correctly the dimmer can overheat and burn / melt, and without calibrations seem very basicly done.

Thank you for the information! I'll check it regulary :-)

**Comment by cjekelund** [comment]
Hoping for Shelly Dimmer 2 support as well, keep up the good work!

**Comment by fernandosierrap** [comment]
no news with that?

**Comment by dreamspy** [comment]
+1 for homekit dimmer support

**Comment by Oshuavid** [comment]
Any update on Dimmer support for Homekit?

**Comment by rojer** [comment]
i'm afraid not

**Comment by tkaudioservices** [comment]
+1 for Dimmer

**Comment by NickSutton** [comment]
+1 for support with Dimmer 2!

**Comment by homkittoby** [comment]
+1 for support with Dimmer 2!

**Comment by homkittoby** [comment]
+1 for support with Dimmer 2! 👌🏻

**Comment by freddair** [comment]
+11 for Dimmer and Dimmer2

**Comment by ActiveByte** [comment]
+1 for support with Dimmer 2!

**Comment by gtudan** [comment]
Please, stop spamming. If you want to express your support for this issue, click +1 on the first comment, or use the comment field in the donation form 😉

**Comment by miguelpruivo** [comment]
To those of you that use dimmer with Shelly homebridge, did you typically experience any issues? Like performance, reliability, flickering etc?

I’m divided whether to pick Shelly 1P or dimmers to my new house but HK support is fundamental here but I also would like to have it work without any kind of issues. At first I suspect that my light bulbs wont be dimmable, does it support traditional lamps as well?

Thank you.

**Comment by miguelpruivo** [comment]
> Hi. Just to inform you. I bought a Shelly Dimmer 2 with Homekit fimrware from this site: https://www.voltio.cz/en/dimmable/548-shelly-dimmer.html .
> It works fine, but it's not possible to calibrate. The lights are flackiring between 30 and 50 %. So the have realized a firmware. Perhaps you can get one to read out it ;-)

Out of curiosity, did someone tried to check how did they do it? Can this be a starting point to help with the implementation for dimmer 2 @rojer? I'm planning to buy the dimmer 2 but won't probably take the risk of buying the voltio ones, despite having HK support.

**Comment by rojer** [comment]
well, their firmware is closed so it's no use for us.
it's not that we don't know what to do, it's just that it's particularly complicated for dimmer compared to other devices (because of the secondary MCU). it's all doable, the only question is time.

**Comment by miguelpruivo** [comment]
Got it @rojer. If I could only have the knowledge that you have regarding this implementation I would probably try to help, but as mainly a mobile developer with just some insights of embedded systems, I’m afraid I can’t help much.

I’ll patiently wait until you come up with something. 💪

**Comment by ratkins** [comment]
[This project](https://github.com/RavenSystem/esp-homekit-devices/wiki) claims Shelly Dimmer support, has anyone tried it?

**Comment by FI97MaWa** [comment]
> Please, stop spamming. If you want to express your support for this issue, click +1 on the first comment, or use the comment field in the donation form 😉

How much should the donation be to have an impact on your motivation 😄
You’re doing a great job!

**Comment by andyblac** [comment]
> > Please, stop spamming. If you want to express your support for this issue, click +1 on the first comment, or use the comment field in the donation form 😉
> 
> How much should the donation be to have an impact on your motivation 😄 You’re doing a great job!

Unfortunately it not about motivation, it about the complexity of the device, the Dimmer2 can over heat and burn if the wrong code is chosen for the wrong bulb (Leading Edge / Trailing Edge), so it requires a lot of work to make it safe to autodetect what bulb type is in use.

@rojer has already blew up his laptop trying to code this device, it had to go in for repair, since then it has been put on back burner as he need to find more time he can allocate to this device.

**Comment by miguelpruivo** [comment]
@andyblac the HAA (Raven esp) has support for it — already tried it and it works nicely but I find the configs a bit complex (seems to be due to the way it optimizes it for performance reasons).

Is this a lot different implementation wise than the HAA?

**Comment by andyblac** [comment]
> @andyblac the HAA (Raven esp) has support for it — already tried it and it works nicely but I find the configs a bit complex (seems to be due to the way it optimizes it for performance reasons).
> 
> Is this a lot different implementation wise than the HAA?
but does it fully support leading / trailing edge or just 'dimming' ?, if you dont have control over what type, then I would not trust it, as the device could overheat and burn, which could cause a fire.

**Comment by FI97MaWa** [comment]
> Unfortunately it not about motivation, it about the complexity of the device, the Dimmer2 can over heat and burn if the wrong code is chosen for the wrong bulb (Leading Edge / Trailing Edge), so it requires a lot of work to make it safe to autodetect what bulb type is in use.


To bad it's not just the motivation. :-) But I understand that it's not that easy. 
Anyway, you guys are doing a great job.

**Comment by robertoamoreno** [comment]
Might help 

Custom firmware for the Shelly Dimmer 2
https://github.com/Mollayo/Shelly-Dimmer-2

Shelly-Dimmer-2-Reverse-Engineering

https://github.com/Mollayo/Shelly-Dimmer-2-Reverse-Engineering

**Comment by Imburdalinix** [comment]
> > > Please, stop spamming. If you want to express your support for this issue, click +1 on the first comment, or use the comment field in the donation form 😉
> 
> > 
> 
> > How much should the donation be to have an impact on your motivation 😄 You’re doing a great job!
> 
> 
> 
> Unfortunately it not about motivation, it about the complexity of the device, the Dimmer2 can over heat and burn if the wrong code is chosen for the wrong bulb (Leading Edge / Trailing Edge), so it requires a lot of work to make it safe to autodetect what bulb type is in use.
> 
> 
> 
> @rojer has already blew up his laptop trying to code this device, it had to go in for repair, since then it has been put on back burner as he need to find more time he can allocate to this device.

Quick question... Can't you just look into the Shelly Source Code and cooy it from there? I am not a programmer or any kind of Tech Pro, but I just wanted to know if this whould work... If the Source Code isn't public it whould make sense that you haven't done it but if it was... then why whouldn't this work? Or whould it?

**Comment by jonathanlang** [comment]
@Imburdalinix If it was that easy, they probably would have done it already... as you can tell from the link above your comment, the firmware has to be reverse engineered and is not just something you can copy & paste.

**Comment by jonathanlang** [comment]
> Since it seems that there's not much chance of any progress soon, does anyone have any suggestions for a HK compatible, safe alternative to shelly dimmer? I need to dim an array of ceiling LED's which I can't swap out to smart GU10's.

I can recommend getting a Raspberry Pi (can be an older model too) and installing Homebridge on it. Then the Pi will act as a basestation and allow control of Shelly dimmers perfectly.

**Comment by just-doit** [comment]
> > Since it seems that there's not much chance of any progress soon, does anyone have any suggestions for a HK compatible, safe alternative to shelly dimmer? I need to dim an array of ceiling LED's which I can't swap out to smart GU10's.
> 
> I can recommend getting a Raspberry Pi (can be an older model too) and installing Homebridge on it. Then the Pi will act as a basestation and allow control of Shelly dimmers perfectly.

I currently have my Shelly Dimmers also connected to Apple Home via a Raspberry Pi with Homebridge. But I would like to do without this "detour".

**Comment by lucassite120** [comment]
I OTA flashed a Shelly Dimmer and it works fine, with https://github.com/RavenSystem/mgostoHAA

**Comment by just-doit** [comment]
> I OTA flashed a Shelly Dimmer and it works fine, with https://github.com/RavenSystem/mgostoHAA

Do you mind to share the MEPLHAA Script you used after flashing the HAA firmware?
